ai8x.py

```diff
@@ -33,6 +33,63 @@ def __call__(self, img):
         return img.sub(0.5).mul(256.).round().clamp(min=-128, max=127).div(128.)
 
 
+class fold:
+    """
+    Fold data to increase the number of channels. An interlaced approach used in this folding
+    as explained in [1].
+
+    [1] https://arxiv.org/pdf/2203.16528.pdf
+    """
+    def __init__(self, fold_ratio):
+        self.fold_ratio = fold_ratio
+
+    def __call__(self, img):
+        if self.fold_ratio == 1:
+            return img
+
+        img_folded = None
+        for i in range(self.fold_ratio):
+            for j in range(self.fold_ratio):
+                img_subsample = img[:, i::self.fold_ratio, j::self.fold_ratio]
+                if img_folded is not None:
+                    img_folded = torch.cat((img_folded, img_subsample), axis=0)
+                else:
+                    img_folded = img_subsample
+
+        return img_folded
+
+
+def unfold_batch(img_batch, fold_ratio):
+    """
+    Unfold data to reduce the number of channels. An interlaced approach used in this folding
+    as explained in [1]. This operation is the reverse of the transformation implemented
+    at ai8x.fold class.
+
+    [1] https://arxiv.org/pdf/2203.16528.pdf
+    """
+    if fold_ratio == 1:
+        return img_batch
+
+    t_device = img_batch.get_device()
+    if t_device < 0:
+        t_device = None
+
+    num_out_channels = img_batch.shape[1] // (fold_ratio*fold_ratio)
+
+    img_batch_uf = torch.zeros((img_batch.shape[0], num_out_channels,
+                                img_batch.shape[2]*fold_ratio, img_batch.shape[3]*fold_ratio),
+                               dtype=img_batch.dtype, device=t_device, requires_grad=False)
```

Comment:
Would getting rid of lines 73-75, and changing line 79 to

img_batch_uf = torch.zeros( ... , device=img_batch.device, ... )

work?

Reply:
This change has been applied. Thanks.